src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights/FeatureManagementBuilderExtensions.cs

```diff
@@ -0,0 +1,40 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.FeatureManagement.Telemetry.ApplicationInsights;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// Extensions used to add feature management functionality.
+    /// </summary>
+    public static class FeatureManagementBuilderExtensions
+    {
+        /// <summary>
+        /// Adds the <see cref="ApplicationInsightsEventPublisher"/> using <see cref="ApplicationInsightsHostedService"/> to the feature management builder.
+        /// </summary>
+        /// <param name="builder">The feature management builder.</param>
+        /// <returns>The feature management builder.</returns>
+        public static IFeatureManagementBuilder AddAppInsightsTelemetryPublisher(this IFeatureManagementBuilder builder)
+        {
+            if (builder == null)
+            {
+                throw new ArgumentNullException(nameof(builder));
+            }
+
+            if (builder.Services == null)
+            {
+                throw new ArgumentNullException(nameof(builder.Services));
+            }
```

Comment:
Not sure what you mean by nested argument validation. `builder` is an argument. The method should ensure that `builder` is in a proper state.

One thing I will say is that `ArgumentNullException` is not the write exception because `builder` is not null in this case. It should be an `ArgumentException` that mentions `builder` is invalid because `services` should be non-null.

Reply:
> Not sure what you mean by nested argument validation. `builder` is an argument. 

`builder` is, but `Services` is not a concern of the method here: it should be ensured/validated when the builder is created. It's more of an SRP thing for me than anything else.

> The method should ensure that `builder` is in a proper state.

Sure, but `builder.Services` is not the argument. If you really want to throw an `ArgumentException`-class exception, it should target `builder` directly because that's the argument.

> One thing I will say is that `ArgumentNullException` is not the write exception because `builder` is not null in this case. It should be an `ArgumentException` that mentions `builder` is invalid because `services` should be non-null.

Agreed. `ArgumentNullException` is misleading here since no argument to the method is `null`.

Reply:
Yes, the paramName argument of `ArgumentException` should be `builder`, not `Services`.